src/transformers/trainer.py

```diff
@@ -811,6 +816,11 @@ def train(
             num_train_epochs = 1
             num_update_steps_per_epoch = max_steps
 
+        if self.args.warmup_ratio > 0 and self.args.warmup_steps == 0:
+            if self.args.warmup_ratio > 1:
+                raise ValueError("warmup_ratio must lie in range [0,1]")
+            self.args.warmup_steps = math.ceil(max_steps * self.args.warmup_ratio)
```

Comment:
We don't change the `args` passed by the user -> if they are re-used for another `Trainer` there will be some surprise on the user side. This test should be made when the optimizer is created, to pass the right value for `warmup_steps`.